### PR TITLE
applications: serial_lte_modem: plug in CMUX to AT commands/responses

### PR DESCRIPTION
**Pull request by tomi-font**
A system of AT backends is introduced to be able to switch from pure UART to CMUX handling of AT data.

One CMUX channel is configured and does the passing of all the AT traffic.

All the normal functionality remains, including going to and resuming from idle mode.

### CODE REVIEW

File: applications/serial_lte_modem/src/slm_uart_handler.c [hunk 1]
```diff
@@ -144,15 +137,13 @@ int slm_uart_rx_disable(void)
 		k_sleep(K_MSEC(10));
 	}
 
-	err = uart_rx_disable(uart_dev);
+	err = uart_rx_disable(slm_uart_dev);
 	if (err) {
 		LOG_ERR("UART RX disable failed: %d", err);
+		atomic_set(&recovery_state, RECOVERY_IDLE);
```
Comment (MarkusLassila) [inline applications/serial_lte_modem/src/slm_uart_handler.c:143]: Unsure if the uart is actually in operationable state in here, but I guess that setting this does not harm.

File: applications/serial_lte_modem/src/slm_cmux.c [hunk 2]
```diff
@@ -5,38 +5,46 @@
...
+	AT_CHANNEL,
+	CHANNEL_COUNT
+};
+
+#define RECV_BUF_LEN SLM_AT_MAX_CMD_LEN
+/* The CMUX module reserves some spare buffer bytes. To achieve a maximum
+ * response length of SLM_AT_MAX_RSP_LEN (comprising the "OK" or "ERROR"
+ * that is sent separately), the transmit buffer must be made a bit bigger.
  */
+#define TRANSMIT_BUF_LEN (49 + SLM_AT_MAX_RSP_LEN)
```
Comment (MarkusLassila) [inline applications/serial_lte_modem/src/slm_cmux.c:27]: Why 49? If there is a reason, you could address that in above comment.

File: applications/serial_lte_modem/src/slm_cmux.c [hunk 3]
```diff
@@ -71,33 +80,84 @@ static void cmux_event_handler(struct modem_cmux *, enum modem_cmux_event event,
...
+	dlci->address = dlci_config.dlci_address;
 
 	modem_pipe_attach(dlci->pipe, dlci_pipe_event_handler, dlci);
 }
 
+static int cmux_write_at_channel(const uint8_t *data, size_t len)
+{
+	struct modem_pipe *const at_dlci_pipe = cmux.dlcis[AT_CHANNEL].pipe;
+	int ret = modem_pipe_transmit(at_dlci_pipe, data, len);
+	const int sent_len = MAX(0, ret);
```
Comment (MarkusLassila) [inline applications/serial_lte_modem/src/slm_cmux.c:105]: Using regular error handling would be easier to follow (first process errors).

File: applications/serial_lte_modem/src/slm_cmux.c [hunk 4]
```diff
@@ -71,33 +80,84 @@ static void cmux_event_handler(struct modem_cmux *, enum modem_cmux_event event,
...
+	}
+	LOG_ERR("Sent %u out of %u AT data bytes. (%d)", sent_len, len, ret);
+	return ret;
+}
+
+static void close_pipe(struct modem_pipe **pipe)
+{
+	if (*pipe) {
+		modem_pipe_close_async(*pipe);
+		modem_pipe_release(*pipe);
```
Comment (MarkusLassila) [inline applications/serial_lte_modem/src/slm_cmux.c:121]: As the above call is asynchronous, what happens if there is more data coming after this?

File: applications/serial_lte_modem/src/slm_cmux.c [hunk 5]
```diff
@@ -71,33 +80,84 @@ static void cmux_event_handler(struct modem_cmux *, enum modem_cmux_event event,
...
-	dlci->address = address;
+	dlci->address = dlci_config.dlci_address;
 
 	modem_pipe_attach(dlci->pipe, dlci_pipe_event_handler, dlci);
 }
 
+static int cmux_write_at_channel(const uint8_t *data, size_t len)
+{
+	struct modem_pipe *const at_dlci_pipe = cmux.dlcis[AT_CHANNEL].pipe;
+	int ret = modem_pipe_transmit(at_dlci_pipe, data, len);
```
Comment (MarkusLassila) [inline applications/serial_lte_modem/src/slm_cmux.c:104]: This buffers the data when the modem pipe is closed without closing cmux in cmux_stop?

File: applications/serial_lte_modem/src/slm_at_host.c [hunk 6]
```diff
@@ -438,6 +442,89 @@ static size_t cmd_name_toupper(char *cmd, size_t cmd_len)
...
+		stop_at_backend();
+
+		at_backend = old_backend;
+		restore_at_backend();
+	}
+
+	return ret;
+}
+
+static int slm_at_send_indicate(const uint8_t *data, size_t len,
```
Comment (MarkusLassila) [inline applications/serial_lte_modem/src/slm_at_host.c:492]: Nice

**Review by SeppoTakalo — APPROVED** [review]

File: applications/serial_lte_modem/src/slm_at_host.c [hunk 7]
```diff
@@ -438,6 +442,89 @@ static size_t cmd_name_toupper(char *cmd, size_t cmd_len)
...
+	const int err = at_backend.stop();
+
+	if (!err) {
+		/* Wait for UART disabling to complete. */
+		k_sleep(K_MSEC(100));
+	}
+	return err;
+}
+
+int slm_at_set_backend(const struct slm_at_backend new_backend)
```
[the remark this answers is not in the document]
Comment (tomi-font) [reply applications/serial_lte_modem/src/slm_at_host.c:466]: You mean from CMUX? This is correct.
I did experiment with this possibility, but faced issues where the application would just hang and so left it out. I don't remember exactly, but possibly found how to do it at some point, so it may be possible to add at a later point if judged useful (currently, `AT#XRESET` does the job).
(Also, I feel like the 3GPP standard for the CMUX command does not define how to stop it. I wonder whether modems supporting CMUX do support stopping it.)

File: applications/serial_lte_modem/src/slm_at_host.c [hunk 8]
```diff
@@ -438,6 +442,89 @@ static size_t cmd_name_toupper(char *cmd, size_t cmd_len)
...
+{
+	const struct slm_at_backend old_backend = at_backend;
+	int ret;
+
+	if (old_backend.start) {
+		ret = stop_at_backend();
```
[the remark this answers is not in the document]
Comment (tomi-font) [reply applications/serial_lte_modem/src/slm_at_host.c:472]: I don't mind doing that, if you think it's better.

File: applications/serial_lte_modem/src/slm_at_host.c [hunk 9]
```diff
@@ -666,45 +745,43 @@ static void notification_handler(const char *notification)
...
-	if (!slm_uart_can_context_send(fmt, strlen(fmt))) {
-		return;
-	}
 	k_mutex_lock(&mutex_rsp_buf, K_FOREVER);
 
 	va_list arg_ptr;
 
 	va_start(arg_ptr, fmt);
-	vsnprintf(rsp_buf, sizeof(rsp_buf), fmt, arg_ptr);
+	rsp_len = vsnprintf(rsp_buf, sizeof(rsp_buf), fmt, arg_ptr);
```
[the remark this answers is not in the document]
Comment (tomi-font) [reply applications/serial_lte_modem/src/slm_at_host.c:774]: Good point. Well spotted.

File: applications/serial_lte_modem/src/slm_cmux.c (hunk 5 above)
Comment (tomi-font) [reply applications/serial_lte_modem/src/slm_cmux.c:105]: Correct.

File: applications/serial_lte_modem/src/slm_cmux.c (hunk 4 above)
Comment (tomi-font) [reply applications/serial_lte_modem/src/slm_cmux.c:123]: This (`close_pipe()`) is only called from `cmux_stop()` when the setup of CMUX has failed (because otherwise only the UART pipe is closed), so there cannot be any data running through the pipes.

Though, in the hypothetical case you are mentioning, the `modem_pipe` module will see that the pipe has been released (`callback` has been set to `NULL`) and simply won't pass events forward (e.g. `MODEM_PIPE_EVENT_RECEIVE_READY`).

File: applications/serial_lte_modem/src/slm_uart_handler.c (hunk 1 above)
Comment (tomi-font) [reply applications/serial_lte_modem/src/slm_uart_handler.c:143]: Yeah, the hypothetical reasoning behind this is that if disabling fails, then it's still running, so set `recovery_state` back to a running state.

File: applications/serial_lte_modem/src/slm_uart_handler.c [hunk 10]
```diff
@@ -450,22 +367,17 @@ int slm_uart_handler_init(slm_uart_rx_callback_t callback_t)
...
-	/* Power off UART module */
-	slm_uart_rx_disable();
-	err = pm_device_action_run(uart_dev, PM_DEVICE_ACTION_SUSPEND);
-	if (err) {
-		LOG_WRN("Can't suspend UART: %d", err);
-	}
+	return slm_at_set_backend((struct slm_at_backend) {
+		.start = slm_uart_handler_init,
+		.send = slm_uart_tx_write,
+		.stop = slm_uart_rx_disable
```
[the remark this answers is not in the document]
Comment (tomi-font) [reply applications/serial_lte_modem/src/slm_uart_handler.c:381]: How about adding a call to `uart_tx_abort()` after the 100ms wait? Would that make sense?

File: applications/serial_lte_modem/src/slm_util.h [hunk 11]
```diff
@@ -21,6 +21,12 @@
 
 extern struct k_work_q slm_work_q; /* SLM's work queue. */
 
+/** @brief Performs a reset of the SiP. */
+FUNC_NORETURN void slm_reset(void);
```
[the remark this answers is not in the document]
Comment (tomi-font) [reply applications/serial_lte_modem/src/slm_util.h:24]: Yeah. (One is actually in `slm_at_commands.c`.) They have some dependencies that should be moved along or declared externally.

File: applications/serial_lte_modem/src/main.c [hunk 12]
```diff
@@ -219,6 +218,7 @@ static void gpio_cb_func(const struct device *dev, struct gpio_callback *gpio_ca
 
 void enter_idle(void)
 {
+	LOG_INF("Entering idle.");
```
[the remark this answers is not in the document]
Comment (MarkusLassila) [reply applications/serial_lte_modem/src/main.c:221]: Hmmh. I guess these are a bit hard to follow without logs. I am fine with this.

File: applications/serial_lte_modem/src/slm_uart_handler.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (MarkusLassila) [reply applications/serial_lte_modem/src/slm_uart_handler.c:381]: It would likely return -EFAULT, which we would need to swallow. Also the uart_callback would just start another TX-operation. So we would need to exit if aborted. And perhaps only if we are closing as it makes sense to retry with other aborts, so this would be bit messy.

Let's add this if we run into problems.

File: applications/serial_lte_modem/src/slm_at_host.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (MarkusLassila) [reply applications/serial_lte_modem/src/slm_at_host.c:472]: We can leave it like this and see how it feels later.

**Review by MarkusLassila — APPROVED** [review]

File: applications/serial_lte_modem/prj.conf [hunk 13]
```diff
@@ -111,6 +111,7 @@ CONFIG_SLM_START_SLEEP=n
 CONFIG_SLM_DATAMODE_URC=n
 
 # Debug configurations
+#CONFIG_LOG_BUFFER_SIZE=16384
```
[the remark this answers is not in the document]
Comment (lemrey) [reply applications/serial_lte_modem/prj.conf:114]: A better option is to collect these values in a dedicated `debug.conf` project configuration overlay file.
Ideally, we shouldn't have commented entries in `prj.conf`.

[the remark this answers is not in the document]
Comment (tomi-font) [reply applications/serial_lte_modem/prj.conf:114]: Seems like a reasonable idea. Though some debugging options are only for certain overlays, but I guess they would just get compiled out as dependencies would be missing. Though that may result in quite some warnings output. Not sure what's the best way.

[the remark this answers is not in the document]
Comment (MarkusLassila) [reply applications/serial_lte_modem/prj.conf:114]: As @tomi-font commented, we have a many overlays with their own debug options. We cannot reasonably expect to have additional overlay files for all the combinations. I would keep this as is for now.
